File: smart contracts/ssv-poc2-multi-cluster/scripts/run_execution_poc.py
"""
Execution PoC: SSV Multi-Cluster Cascading Insolvency

This script simulates the multi-cluster insolvency attack,
demonstrating how multiple bankrupt clusters compound the theft.
"""

def run_execution_poc():
    print("=" * 80)
    print("SSV MULTI-CLUSTER CASCADING INSOLVENCY: EXECUTION TRACE")
    print("=" * 80)

    # 1. Initial State - Multiple Clusters
    pool_assets = 0
    
    # Large user deposits
    large_deposit = 10000
    pool_assets += large_deposit
    
    # Small users deposit (will go bankrupt)
    small_1 = 100  # Bankrupts in 100 blocks
    small_2 = 50   # Bankrupts in 50 blocks
    small_3 = 25   # Bankrupts in 25 blocks
    pool_assets += small_1 + small_2 + small_3
    
    print("Block 0 - Initial Deposits:")
    print("  Large User:    " + str(large_deposit) + " SSV")
    print("  Small User 1:  " + str(small_1) + " SSV")
    print("  Small User 2:  " + str(small_2) + " SSV")
    print("  Small User 3:  " + str(small_3) + " SSV")
    print("  Total Assets:  " + str(pool_assets) + " SSV")

    # 2. Setup Operators
    op_fee = 1  # 1 SSV per block per operator
    
    # 3. Simulate 150 blocks
    current_block = 150
    print("\n--- 150 Blocks Pass ---")
    
    # Cluster 1 (Small 1): Bankrupt at block 100
    # Virtual debt for 50 blocks = 50 SSV
    virtual_debt_1 = 50 * op_fee
    print("Block 150 - Small Cluster 1: BANKRUPT")
    print("  Virtual debt to Operator 1: " + str(virtual_debt_1) + " SSV")
    
    # Cluster 2 (Small 2): Bankrupt at block 50
    # Virtual debt for 100 blocks = 100 SSV
    virtual_debt_2 = 100 * op_fee
    print("Block 150 - Small Cluster 2: BANKRUPT")
    print("  Virtual debt to Operator 2: " + str(virtual_debt_2) + " SSV")
    
    # Cluster 3 (Small 3): Bankrupt at block 25
    # Virtual debt for 125 blocks = 125 SSV
    virtual_debt_3 = 125 * op_fee
    print("Block 150 - Small Cluster 3: BANKRUPT")
    print("  Virtual debt to Operator 3: " + str(virtual_debt_3) + " SSV")
    
    # DAO fees from all clusters (0.5 SSV per block per validator)
    dao_virtual_debt = 150 * 0.5 * 3  # 225 SSV
    print("  DAO unbacked fees:           " + str(dao_virtual_debt) + " SSV")
    
    total_virtual_debt = virtual_debt_1 + virtual_debt_2 + virtual_debt_3 + dao_virtual_debt
    print("\nTOTAL VIRTUAL DEBT: " + str(total_virtual_debt) + " SSV")
    print("(This debt is UNBACKED - clusters have no funds to pay it)")
    
    # 4. Bank Run - Operators race to withdraw
    print("\n--- BANK RUN: Race to Withdraw ---")
    
    # Operator 3 withdraws first
    withdrawal_3 = virtual_debt_3
    if withdrawal_3 <= pool_assets:
        pool_assets -= withdrawal_3
        print("Operator 3 withdrew:  " + str(withdrawal_3) + " SSV")
    
    # Operator 2 withdraws second
    withdrawal_2 = virtual_debt_2
    if withdrawal_2 <= pool_assets:
        pool_assets -= withdrawal_2
        print("Operator 2 withdrew:  " + str(withdrawal_2) + " SSV")
    
    # Operator 1 withdraws third
    withdrawal_1 = virtual_debt_1
    if withdrawal_1 <= pool_assets:
        pool_assets -= withdrawal_1
        print("Operator 1 withdrew:  " + str(withdrawal_1) + " SSV")
    
    # DAO withdraws
    dao_withdrawal = dao_virtual_debt
    if dao_withdrawal <= pool_assets:
        pool_assets -= dao_withdrawal
        print("DAO withdrew:         " + str(dao_withdrawal) + " SSV")
    
    total_stolen = withdrawal_1 + withdrawal_2 + withdrawal_3 + dao_withdrawal
    print("\nTotal stolen:         " + str(total_stolen) + " SSV")

    # 5. Honest victim attempts withdrawal
    print("\n--- Honest Large User Attempts Withdrawal ---")
    print("Large user entitlement: " + str(large_deposit) + " SSV")
    print("Remaining pool assets:  " + str(pool_assets) + " SSV")
    
    if large_deposit <= pool_assets:
        print("SUCCESS: Large user recovered all funds.")
        loss = 0
    else:
        loss = large_deposit - pool_assets
        print("CRITICAL FAILURE: Large user can only withdraw " + str(pool_assets) + " SSV")
        print("LARGE USER TOTAL LOSS: " + str(loss) + " SSV")

    print("\n" + "=" * 80)
    print("CONCLUSION: Multi-Cluster Cascading Insolvency Proven.")
    print("Three bankrupt clusters created " + str(total_virtual_debt) + " SSV of virtual debt.")
    print("This demonstrates systemic risk - each additional cluster compounds the theft.")
    print("=" * 80)

if __name__ == "__main__":
    run_execution_poc()

File: smart contracts/ssv-poc2-multi-cluster/scripts/old/run-execution-poc.js
/**
 * Execution PoC: SSV Multi-Cluster Cascading Insolvency
 * 
 * This script simulates the multi-cluster insolvency attack,
 * demonstrating how multiple bankrupt clusters compound the theft.
 * 
 * Run with: node scripts/run-execution-poc.js
 */

function runExecutionPoC() {
    console.log("=".repeat(80));
    console.log("SSV MULTI-CLUSTER CASCADING INSOLVENCY: EXECUTION TRACE");
    console.log("=".repeat(80));

    // 1. Initial State - Multiple Clusters
    let poolAssets = 0;
    
    // Large user deposits
    const largeDeposit = 10000;
    const small1 = 100;  // Bankrupts in 100 blocks
    const small2 = 50;   // Bankrupts in 50 blocks
    const small3 = 25;   // Bankrupts in 25 blocks
    poolAssets += largeDeposit + small1 + small2 + small3;
    
    console.log("Block 0 - Initial Deposits:");
    console.log(`  Large User:    ${largeDeposit} SSV`);
    console.log(`  Small User 1:  ${small1} SSV`);
    console.log(`  Small User 2:  ${small2} SSV`);
    console.log(`  Small User 3:  ${small3} SSV`);
    console.log(`  Total Assets:  ${poolAssets} SSV`);

    // 2. Setup Operators
    const opFee = 1;  // 1 SSV per block per operator
    
    // 3. Simulate 150 blocks
    const currentBlock = 150;
    console.log("\n--- 150 Blocks Pass ---");
    
    // Cluster 1 (Small 1): Bankrupt at block 100
    // Virtual debt for 50 blocks = 50 SSV
    const virtualDebt1 = 50 * opFee;
    console.log(`Block 150 - Small Cluster 1: BANKRUPT`);
    console.log(`  Virtual debt to Operator 1: ${virtualDebt1} SSV`);
    
    // Cluster 2 (Small 2): Bankrupt at block 50
    // Virtual debt for 100 blocks = 100 SSV
    const virtualDebt2 = 100 * opFee;
    console.log(`Block 150 - Small Cluster 2: BANKRUPT`);
    console.log(`  Virtual debt to Operator 2: ${virtualDebt2} SSV`);
    
    // Cluster 3 (Small 3): Bankrupt at block 25
    // Virtual debt for 125 blocks = 125 SSV
    const virtualDebt3 = 125 * opFee;
    console.log(`Block 150 - Small Cluster 3: BANKRUPT`);
    console.log(`  Virtual debt to Operator 3: ${virtualDebt3} SSV`);
    
    // DAO fees from all clusters (0.5 SSV per block per validator)
    const daoVirtualDebt = 150 * 0.5 * 3;  // 225 SSV
    console.log(`  DAO unbacked fees:           ${daoVirtualDebt} SSV`);
    
    const totalVirtualDebt = virtualDebt1 + virtualDebt2 + virtualDebt3 + daoVirtualDebt;
    console.log(`\nTOTAL VIRTUAL DEBT: ${totalVirtualDebt} SSV`);
    console.log("(This debt is UNBACKED - clusters have no funds to pay it)");
    
    // 4. Bank Run - Operators race to withdraw
    console.log("\n--- BANK RUN: Race to Withdraw ---");
    
    // Operator 3 withdraws first
    const withdrawal3 = virtualDebt3;
    if (withdrawal3 <= poolAssets) {
        poolAssets -= withdrawal3;
        console.log(`Operator 3 withdrew:  ${withdrawal3} SSV`);
    }
    
    // Operator 2 withdraws second
    const withdrawal2 = virtualDebt2;
    if (withdrawal2 <= poolAssets) {
        poolAssets -= withdrawal2;
        console.log(`Operator 2 withdrew:  ${withdrawal2} SSV`);
    }
    
    // Operator 1 withdraws third
    const withdrawal1 = virtualDebt1;
    if (withdrawal1 <= poolAssets) {
        poolAssets -= withdrawal1;
        console.log(`Operator 1 withdrew:  ${withdrawal1} SSV`);
    }
    
    // DAO withdraws
    const daoWithdrawal = daoVirtualDebt;
    if (daoWithdrawal <= poolAssets) {
        poolAssets -= daoWithdrawal;
        console.log(`DAO withdrew:         ${daoWithdrawal} SSV`);
    }
    
    const totalStolen = withdrawal1 + withdrawal2 + withdrawal3 + daoWithdrawal;
    console.log(`\nTotal stolen:         ${totalStolen} SSV`);
    console.log("All of it is UNBACKED virtual debt!");

    // 5. Honest victim attempts withdrawal
    console.log("\n--- Honest Large User Attempts Withdrawal ---");
    console.log(`Large user entitlement: ${largeDeposit} SSV`);
    console.log(`Remaining pool assets:  ${poolAssets} SSV`);
    
    if (largeDeposit <= poolAssets) {
        console.log("SUCCESS: Large user recovered all funds.");
    } else {
        const loss = largeDeposit - poolAssets;
        console.log(`CRITICAL FAILURE: Large user can only withdraw ${poolAssets} SSV`);
        console.log(`LARGE USER TOTAL LOSS: ${loss} SSV`);
    }

    console.log("\n" + "=".repeat(80));
    console.log("CONCLUSION: Multi-Cluster Cascading Insolvency Proven.");
    console.log(`Three bankrupt clusters created ${totalVirtualDebt} SSV of virtual debt.`);
    console.log("This demonstrates systemic risk - each additional cluster compounds the theft.");
    console.log("=".repeat(80));
}

// Run the PoC
runExecutionPoC();

module.exports = { runExecutionPoC };
